### ISSUE TITLE
Transactions: Sold stock not removed from transaction

### ISSUE THREAD

**Issue by rawcreative** [stevebauman/inventory#16]
While going through and checking out v1.5 I noticed that any transaction that is marked as sold is not removed from the transaction. If someone happens to be comparing the quantity_after column on sold transactions to something, all counts will be incorrect. Since the stock is being released, this should be removed from the transaction when it's marked as sold. It technically should create a stock movement but I don't think this is possible since it's already been reduced with the reservation.

**Comment by stevebauman** [comment]
Hmm... Are you saying that when a transaction is marked sold, the quantity isn't reset to `0` on the transaction itself? If this is the case then this is intended due to the potential of the `returned()` method being called on the transaction to be able to return the sold quantity (the quantity that remains on the transaction) to the stock. Is that what you're referring to?

**Comment by rawcreative** [comment]
No, not the quantity on the transaction, the quantity_after field on transaction history record. It’s not adjusted to reflect that the stock was released from being reserved, i.e. the quantity_before and quantity_after amounts are the same.

**Comment by rawcreative** [comment]
As a side note, a metric I’d like to track is time from sale to ship. When the sale is completed it’s marked as reserved/checkout, when it’s shipped it’s marked as sold. If the transaction contains a qty of two, it would be nice to be able to mark 1 as sold to note that it’s been released, but the way it’s set up you currently can’t do that because you can only mark new transactions as sold.

**Comment by stevebauman** [comment]
This was intended but now I definitely understand what you're saying. The before and after quantity were the same because technically the quantity inside the transaction record isn't changing going from a reserved to sold like so:

    $transaction->reserved(5)->sold();

When the transaction is marked sold from being reserved it doesn't modify the quantity since the reserved function has already taken out 5 quantity from the stock, so the sold function needs the quantity to be able to perform a return through `returned()`. So the transaction history would look like this:

<table class="tg">
  <tr>
    <th class="tg-031e">state_before</th>
    <th class="tg-031e">state_after</th>
    <th class="tg-031e">quantity_before</th>
    <th class="tg-031e">quantity_after</th>
  </tr>
  <tr>
    <td class="tg-031e">NULL</td>
    <td class="tg-031e">opened</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">0</td>
  </tr>
  <tr>
    <td class="tg-031e">opened</td>
    <td class="tg-031e">commerce-reserved</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">5</td>
  </tr>
  <tr>
    <td class="tg-031e">commerce-reserved</td>
    <td class="tg-031e">commerce-sold</td>
    <td class="tg-031e">5</td>
    <td class="tg-031e">5</td>
  </tr>
</table>

This is because the quantity isn't changing, so the history is correct. The quantity before the change was `5` and still is `5`. Just to clarify, this is what you're mentioning right?

**Comment by rawcreative** [comment]
Yeah, that’s what I’m referring to. I see what is happening and understand that a reference needs to be held for returns sake, but based on how all other states handle the quantity_after column, it should really be 0 since the stock is no longer anywhere in inventory and has in fact been completely released. I can see both sides though, where that column could be used as a running tally of what’s sold, but it seems out of place since all other states increment or decrement that field based on the status of the item in inventory. I think the argument could be made that both ways are correct, it’s just a matter of how they are being applied in a given use case. 

I think for my particular use cause, I’m probably going to have to add an additional ‘commerce-shipped’ state to be able to track the metrics I want to track, and then funnel the sold state into that.

**Comment by stevebauman** [comment]
Yea you're right. This needs to be changed to reflect quantity that has actually left the stock, and the `sold()` method should perform this. 

I'm not too sure how to go about this though since I'd rather not have to dig into the transaction history to be able to perform a return. However, maybe a `returned()` transaction can be modified to be a whole new transaction? Thoughts?

**Comment by rawcreative** [comment]
I think that’s perfectly acceptable to have a separate transaction for returns. I would say that in most instances returned stock is treated differently than regular stock anyway since it is not always in a ‘new’ condition, or may be defective, etc. So it could be helpful to separate them for that reason.

**Comment by stevebauman** [comment]
That's very true. Ok, returns will be modified so they are only accessible through a new stock transaction. Also, here's a new transaction history schema I'm suggesting, definitely let me know on your thoughts. This is referencing the transaction mentioned above:

<table class="tg">
  <tr>
    <th class="tg-031e">state_before</th>
    <th class="tg-031e">state_after</th>
    <th class="tg-031e">quantity_held_before</th>
    <th class="tg-031e">quantity_held_after</th>
    <th class="tg-031e">quantity_removed_before</th>
    <th class="tg-031e">quantity_removed_after</th>
  </tr>
  <tr>
    <td class="tg-031e">NULL</td>
    <td class="tg-031e">opened</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">0</td>
  </tr>
  <tr>
    <td class="tg-031e">opened</td>
    <td class="tg-031e">commerce-reserved</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">5</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">0</td>
  </tr>
  <tr>
    <td class="tg-031e">commerce-reserved</td>
    <td class="tg-031e">commerce-sold</td>
    <td class="tg-031e">5</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">5</td>
  </tr>
</table>

This schema would correctly show that a reservation took place, and the quantity was removed and held from the stock. Then it was sold and was removed completely from the stock. Not sure on the column names, but possible a start.

Also, the transactions table would get a new column `quantity_held` and the `quantity` column would be renamed to `quantity_removed`. Migrations would be supplied to any users can easily update to it as well.

**Comment by rawcreative** [comment]
My thoughts on the schema are that it might be overkill to separate everything like that. I think it's possible to allow for everything to be accomplished as it should just by adding another column titled initial_quantity to transactions.

So for transactions the initial_quantity would be set on creation and then never changed and the quantity column updated based on state. This would allow the histories to correctly reflect before and after quantities without reworking everything, since ultimately it's just the sold state that isn't reflected properly.

Thoughts?

**Comment by stevebauman** [comment]
Hmm, adding an `initial_quantity` column seems like it could add some confusion here unless I'm not understanding which could be the case. The column sounds like it should be kept in the transaction history since it is by definition holding a value that was created in the past. For example:

<table class="tg">
  <tr>
    <th class="tg-031e">state</th>
    <th class="tg-031e">initial_quantity</th>
    <th class="tg-031e">quantity</th>
  </tr>
  <tr>
    <td class="tg-031e">commerce-reserved</td>
    <td class="tg-031e">5</td>
    <td class="tg-031e">0</td>
  </tr>
</table>

Taking a look at the table, would both of these methods produce the record above?:

    $transaction->checkout(5)->reserved();

    // Or

    $transaction->reserved(5);

Or would calling:

    $transaction->checkout(5)->reserved();

Result in the transaction record:

<table class="tg">
  <tr>
    <th class="tg-031e">state</th>
    <th class="tg-031e">initial_quantity</th>
    <th class="tg-031e">quantity</th>
  </tr>
  <tr>
    <td class="tg-031e">commerce-reserved</td>
    <td class="tg-031e">5</td>
    <td class="tg-031e">5</td>
  </tr>
</table>

There's also the issue of wanting to know the history behind the `initial_quantity`. Users might wonder: "What was the state of the transaction when `initial_quantity` was set?". Since no history is generated on the `initial_quantity`, we don't know besides going into the history and assuming it was the first state of the transaction.

With this method, all we know is the current state, and the `initial_quantity` which may have or may not have been set <i>with</i> the current state.

However, naming the columns differently:

<table class="tg">
  <tr>
    <th class="tg-031e">state</th>
    <th class="tg-031e">quantity_held</th>
    <th class="tg-031e">quantity_removed</th>
  </tr>
  <tr>
    <td class="tg-031e">commerce-reserved</td>
    <td class="tg-031e">5</td>
    <td class="tg-031e">0</td>
  </tr>
</table>

We know right now at it's current state that it is currently holding 5 stock for a reservation.

And once the reservation has been sold, it would produce:

<table class="tg">
  <tr>
    <th class="tg-031e">state</th>
    <th class="tg-031e">quantity_held</th>
    <th class="tg-031e">quantity_removed</th>
  </tr>
  <tr>
    <td class="tg-031e">commerce-sold</td>
    <td class="tg-031e">0</td>
    <td class="tg-031e">5</td>
  </tr>
</table>

Now we know that this transaction has permanently removed the stock from the inventory because it's current state is sold.

Inside the transaction history, all of the values are traceable in the history, allowing you to see the modifications in both columns when a state has been changed.

**Comment by rawcreative** [comment]
I see where you're coming from with the new columns, and it makes sense. Initially I think I didn't realize that those columns were only pertaining to the current status in the history, not a 'running' history as it is now with the before/after. My thought with the initial_quantity is that it would be set with whatever amounts were initially passed in the transaction but I forgot to take into account that there's no real history of state on the transaction itself so that column quickly becomes meaningless when the state changes.

**Comment by stevebauman** [comment]
Ok, I'll work on this tonight and tomorrow. I'll propose the changes on here before I push them to the master branch so I can make sure I'm going in the right direction. Once this gets released I'll continue on assemblies. I appreciate your input!
